AsyncDisplayKit/ASImageNode.mm

```diff
@@ -410,6 +415,174 @@ - (void)setImageModificationBlock:(asimagenode_modification_block_t)imageModific
+      yaw = (point.x/CGRectGetWidth(self.view.frame))*10;
+    }
+    if (!topHalf) {
+      if (yaw > 0) {
+        yaw = -yaw;
+      } else {
+        yaw = fabsf(yaw);
+      }
+    }
+
+    CATransform3D pitchTransform = CATransform3DMakeRotation([self degressToRadians:pitch],1.0,0.0,0.0);
+    CATransform3D yawTransform = CATransform3DMakeRotation([self degressToRadians:yaw],0.0,1.0,0.0);
+    CATransform3D transform = CATransform3DConcat(pitchTransform, yawTransform);
+    CATransform3D scaleAndTransform = CATransform3DConcat(transform, CATransform3DMakeAffineTransform(CGAffineTransformScale(CGAffineTransformIdentity, 1.25, 1.25)));
+    
+    [UIView animateWithDuration:0.5 animations:^{
+      view.layer.transform = scaleAndTransform;
+    }];
+  } else {
+    [self setDefaultState];
+  }
+}
+
+
+- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
+{
+  [super touchesEnded:touches withEvent:event];
+  [self finishTouches];
+}
+
+- (void)finishTouches
+{
+  if (!isDefaultState) {
+    UIView *view = [self getView];
+    CALayer *layer = view.layer;
+    
+    CGSize targetShadowOffset = CGSizeMake(0.0, self.bounds.size.height/8);
+    CATransform3D targetScaleTransform = CATransform3DMakeScale(1.2, 1.2, 1.2);
+    [CATransaction begin];
+    [CATransaction setCompletionBlock:^{
+      layer.shadowOffset = targetShadowOffset;
+    }];
+    [CATransaction commit];
+    
+    [UIView animateWithDuration:0.4 delay:0.0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
+      view.layer.transform = targetScaleTransform;
+    } completion:^(BOOL finished) {
+      if (finished) {
+        [layer removeAnimationForKey:@"shadowOffset"];
+        [layer removeAnimationForKey:@"shadowOpacity"];
+      }
+    }];
+  } else {
+    [self setDefaultState];
+  }
+}
+
+- (void)setFocusedState
+{
+  UIView *view = [self getView];
+  CALayer *layer = view.layer;
+  layer.shadowOffset = CGSizeMake(2, 10);
+  layer.shadowColor = [UIColor blackColor].CGColor;
+  layer.shadowRadius = 12.0;
+  layer.shadowOpacity = 0.45;
+  layer.shadowPath = [UIBezierPath bezierPathWithRect:self.layer.bounds].CGPath;
+  view.transform = CGAffineTransformScale(CGAffineTransformIdentity, 1.25, 1.25);
+}
+
+- (void)setDefaultState
+{
+  UIView *view = [self getView];
+  CALayer *layer = view.layer;
+  view.transform = CGAffineTransformIdentity;
+  layer.shadowOpacity = 0;
+  layer.shadowOffset = CGSizeZero;
+  layer.shadowRadius = 0;
+  layer.shadowPath = nil;
+  [layer removeAnimationForKey:@"shadowOffset"];
+  [layer removeAnimationForKey:@"shadowOpacity"];
+  isDefaultState = YES;
+}
+
+- (UIView *)getView
+{
+  UIView *view = self.view;
+  //If we are inside a ASCellNode, then we need to apply our focus effects to the ASCellNode view/layer rather than the ASImageNode view/layer.
+  if (CGSizeEqualToSize(self.view.superview.frame.size, self.view.frame.size) && self.view.superview.superview) {
+    view = self.view.superview.superview;
```

Comment:
Hmm...this isn't the right way to do this.  If we need to walk parents to find a cell, that's one thing.  We can check .hierarchyState for rangeManaged, which means that we're in an ASCellNode.  But what if we're inside another view deep?

There is a Helper function somewhere for the first ancestor of nodeclass or something.